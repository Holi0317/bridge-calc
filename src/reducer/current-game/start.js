// @flow
import {GameStage} from '../../game-stage'
import {skip} from './skip'

import type {START_ACTION} from '../../actions/current-game'
import type {WaitingBidState} from './types'

/**
 * Helper function for reducer.
 * Start a new game base on start action parameters.
 * @param action
 */
export function start(action: START_ACTION) {
  // Create an array of playerID -> empty array
  const scores_ = Object.keys(action.playerNames)
    .map(id => ({[id]: []}))

  const firstState: WaitingBidState = {
    stage: GameStage.waitingBid,
    rounds: action.rounds,
    startTime: action.startTime,
    names: action.playerNames,
    scores: Object.assign({}, ...scores_),
    bid: {},
    currentPlayerOrder: Object.keys(action.playerNames),
    currentRound: 1
  }
  return skip(firstState, action.startingRound - 1, action.startTime)
}
