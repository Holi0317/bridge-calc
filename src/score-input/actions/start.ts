import cuid from 'cuid'
import {IPlayerMap} from '../../types'
import {ActionTypes} from '../../action-types'

export interface IStartAction {
  type: ActionTypes.START,
  rounds: number,
  playerNames: IPlayerMap<string>,
  startTime: number,
  startingRound: number,
  id: string
}

/**
 * Helper function to Change player names array to object with random generated player ID as key.
 * @param playerNames
 */
function namesToMap(playerNames: string[]): IPlayerMap<string> {
  const result: IPlayerMap<string> = {}
  playerNames.forEach(name => {
    result[cuid()] = name
  })
  return result
}

/**
 * Start a new game that overrides existing game.
 *
 * @param rounds - Total number of rounds in the whole game.
 * @param playerNames - Player names in the game. Each player will get an unique ID when game starts.
 * Ordering of the names will be used as the player order for first round.
 * @param startingRound - The round to start from
 */
export function startAction(rounds: number, playerNames: string[], startingRound: number): IStartAction {
  return {
    type: ActionTypes.START,
    rounds,
    playerNames: namesToMap(playerNames),
    startTime: new Date().getTime(),
    startingRound,
    id: cuid()
  }
}
