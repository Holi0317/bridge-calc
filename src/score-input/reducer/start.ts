import {GameStage} from '../game-stage'
import {skip} from './skip'
import {bidWinGenerator} from './bid-win-generator'
import {fillObj} from '../../utils'

import {IStartAction} from '../../actions/current-game'
import {IWaitingBidState} from './types'

/**
 * Helper function for reducer.
 * Start a new game base on start action parameters.
 * @param action
 */
export function start(action: IStartAction) {
  const playerIDs = Object.keys(action.playerNames)

  const firstState: IWaitingBidState = {
    stage: GameStage.waitingBid,
    rounds: action.rounds,
    startTime: action.startTime,
    names: action.playerNames,
    scores: fillObj({}, playerIDs, []),
    bid: bidWinGenerator(playerIDs),
    currentPlayerOrder: playerIDs,
    currentRound: 1
  }
  return skip(firstState, action.startingRound - 1, action.startTime)
}
