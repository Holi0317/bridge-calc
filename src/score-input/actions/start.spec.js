import {startAction} from './start'
import * as lolex from 'lolex'

jest.mock('cuid', () => {
  let count = 0
  return jest.fn(() => '' + count++)
})

test('it should returns start action', () => {
  const clock = lolex.install()
  const actual = startAction(13, ['John', 'Mary', 'Henry', 'Joe'], 1)
  expect(actual).toMatchSnapshot()
  clock.uninstall()
})
