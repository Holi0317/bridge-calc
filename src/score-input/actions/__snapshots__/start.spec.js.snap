// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it should returns start action 1`] = `
Object {
  "id": 4,
  "playerNames": Object {
    "0": "John",
    "1": "Mary",
    "2": "Henry",
    "3": "Joe",
  },
  "rounds": 13,
  "startTime": 1970-01-01T00:00:00.000Z,
  "startingRound": 1,
  "type": "CURRENT_GAME/START",
}
`;
